### PR TITLE
feat: add unmount field validation logic to JSON form widget (#37220)

### LINKED ISSUE #18752: [Bug]-[72]:Submit gets unclickable after deleting all array item from JSONForm
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Description

https://www.loom.com/share/70b55b90403c4f10aefb15f1adb1799c

### Steps To Reproduce

1 Add JSON form onto a canvas
2 Add a new field as Array and let the field be a required field
3 Click on 'Add New' in the form to add more array items and let them be empty
4 Now delete all the entries and Observe that submit is not clickable

### Public Sample App

_No response_

### Issue video log

_No response_

### Version

1.8.3

### PR DESCRIPTION
**Pull request by rahulbarwal**
## Description
<ins>Problem</ins>

When deleting all fields of array item, submit became disabled.

<ins>Root cause</ins>

The JSON form widget did not properly handle field validation during unmounting of array items, leading to inconsistencies in the form's error state.

<ins>Solution</ins>

This PR implements `useUnmountFieldValidation` to `FieldRenderer.tsx`, enhancing field validation for array items in the JSON form widget. This PR handles... 

- Ensuring proper cleanup of field validation when an array field is removed from the form, resolving visibility issues in list view mode.
- Maintaining a consistent and accurate form state even after field removal.

Fixes #18752
_or_  
Fixes `Issue URL`
> [!WARNING]  
> _If no issue exists, please create an issue first, and check with the maintainers if the issue is valid._

## Automation

/ok-to-test tags="@tag.JSONForm"

### :mag: Cypress test results

> [!TIP]
> 🟢 🟢 🟢 All cypress tests have passed! 🎉 🎉 🎉
> Workflow run: <https://github.com/appsmithorg/appsmith/actions/runs/11703351095>
> Commit: 521fd25a83c099ddcc0273cf11b4cca80074c7a5
> <a href="https://internal.appsmith.com/app/cypress-dashboard/rundetails-65890b3c81d7400d08fa9ee5?branch=master&workflowId=11703351095&attempt=1" target="_blank">Cypress dashboard</a>.
> Tags: `@tag.JSONForm`
> Spec:
> <hr>Wed, 06 Nov 2024 12:45:42 UTC

## Communication
Should the DevRel and Marketing teams inform users about this change?
- [ ] Yes
- [ ] No

## Summary by CodeRabbit

- **New Features**
	- Introduced a new mechanism for handling field validation upon unmounting in the form widget.
  
- **Bug Fixes**
	- Improved field validation logic to enhance form data management during component unmounting.

### CODE REVIEW

File: app/client/src/widgets/JSONFormWidget/fields/FieldRenderer.tsx [hunk 1]
```diff
@@ -30,6 +31,7 @@ function FieldRenderer({
 
   const FieldComponent = FIELD_MAP[fieldType];
 
+  useUnmountFieldValidation({ fieldName });
```
Comment (ashit-rath) [inline app/client/src/widgets/JSONFormWidget/fields/FieldRenderer.tsx:34]: If you are adding it here; then doesn't it mean we might not need to explicitly do it for every field?

Comment (rahulbarwal) [reply app/client/src/widgets/JSONFormWidget/fields/FieldRenderer.tsx:34]: Basically this problem #28018 appears again if we remove it from all the other places.

Comment (ashit-rath) [reply app/client/src/widgets/JSONFormWidget/fields/FieldRenderer.tsx:34]: Got it

**Review by ashit-rath — APPROVED** [review]

**Pull request merged**